fdbserver/KeyValueStoreShardedRocksDB.actor.cpp

```diff
@@ -2220,64 +2294,141 @@ struct ShardedRocksDBKeyValueStore : IKeyValueStore {
-				return;
-			}
+			TraceEvent(SevDebug, "ShardedRocksCheckpointCF", logId)
+			    .detail("CFName", ps->cf->GetName())
+			    .detail("CheckpointID", a.request.checkpointID)
+			    .detail("Version", a.request.version)
+			    .detail("Ranges", describe(a.request.ranges))
+			    .detail("PhysicalShardRanges", describe(ps->getAllRanges()))
+			    .detail("Format", static_cast<int>(a.request.format))
+			    .detail("CheckpointDir", a.request.checkpointDir);
 
 			rocksdb::PinnableSlice value;
 			rocksdb::ReadOptions readOptions = getReadOptions();
 			s = a.shardManager->getDb()->Get(
 			    readOptions, a.shardManager->getSpecialKeysShard()->cf, toSlice(persistVersion), &value);
 
 			if (!s.ok() && !s.IsNotFound()) {
-				logRocksDBError(s, "Checkpoint");
-				a.reply.sendError(statusToError(s));
+				logRocksDBError(s, "ShardedRocksCheckpointReadPersistVersion");
+				a.reply.sendError(failed_to_create_checkpoint());
 				return;
 			}
 
 			const Version version = s.IsNotFound()
 			                            ? latestVersion
 			                            : BinaryReader::fromStringRef<Version>(toStringRef(value), Unversioned());
 
-			TraceEvent(SevDebug, "ShardedRocksDBServeCheckpointVersion", logId)
+			TraceEvent(SevDebug, "ShardedRocksCheckpointVersion", logId)
 			    .detail("CheckpointVersion", a.request.version)
 			    .detail("PersistVersion", version);
 			ASSERT(a.request.version == version || a.request.version == latestVersion);
 
 			CheckpointMetaData res(ps->getAllRanges(), version, a.request.format, a.request.checkpointID);
-			const std::string& checkpointDir = abspath(a.request.checkpointDir);
+			s = rocksdb::Checkpoint::Create(a.shardManager->getDb(), &checkpoint);
+			if (!s.ok()) {
+				logRocksDBError(s, "CreateRocksDBCheckpoint");
+				a.reply.sendError(failed_to_create_checkpoint());
+				return;
+			}
 
+			const std::string& checkpointDir = abspath(a.request.checkpointDir);
+			platform::eraseDirectoryRecursive(checkpointDir);
 			if (a.request.format == DataMoveRocksCF) {
-				rocksdb::ExportImportFilesMetaData* pMetadata;
-				platform::eraseDirectoryRecursive(checkpointDir);
+				rocksdb::ExportImportFilesMetaData* pMetadata = nullptr;
 				s = checkpoint->ExportColumnFamily(ps->cf, checkpointDir, &pMetadata);
 				if (!s.ok()) {
-					logRocksDBError(s, "ExportColumnFamily");
-					a.reply.sendError(statusToError(s));
+					logRocksDBError(s, "CheckpointExportColumnFamily");
+					a.reply.sendError(failed_to_create_checkpoint());
 					return;
 				}
 
-				populateMetaData(&res, *pMetadata);
+				populateMetaData(&res, pMetadata);
+				rocksdb::ExportImportFilesMetaData metadata = *pMetadata;
 				delete pMetadata;
-				TraceEvent(SevInfo, "ShardedRocksDBServeCheckpointSuccess", logId)
-				    .detail("CheckpointMetaData", res.toString())
-				    .detail("RocksDBCF", getRocksCF(res).toString());
+				if (SERVER_KNOBS->ROCKSDB_ENABLE_CHECKPOINT_VALIDATION) {
```

Comment:
Is this only for testing?

Reply:
Yes. For both sim-test and other real cluster tests.